reccmp/tools/decomplint.py

```diff
@@ -90,35 +100,61 @@ def process_files(files, module=None):
         warning_count += len(warnings)
 
         if not success:
-            display_errors(linter.alerts, filename)
+            display_errors(linter.alerts, file.path)
             print()
 
     return (warning_count, error_count)
 
 
 def main():
     args = parse_args()
+    files_to_check: list[Path] = []
 
     if not args.paths:
+        # No path specified. Try to find the project file.
         project = RecCmpProject.from_directory(directory=Path.cwd())
         if not project:
             logger.error("Cannot find reccmp project")
             return 1
-        print(project.targets)
-        args.paths = list(
-            set(target.source_root for target in project.targets.values())
-        )
-
-    files_to_check: list[str] = []
-    for path in args.paths:
-        if path.is_dir():
-            files_to_check.extend(walk_source_dir(path))
-        elif path.is_file() and is_file_c_like(path):
-            files_to_check.append(str(path))
-        else:
-            logger.error("Invalid path: %s", path)
 
-    (warning_count, error_count) = process_files(files_to_check, module=args.module)
+        # Read each target from the reccmp-project file
+        # then get all filenames from each code directory.
+        for target in project.targets.values():
+            if target.source_root:
+                files_to_check.extend(walk_source_dir(target.source_root))
+    else:
+        for path in args.paths:
+            if path.is_dir():
+                files_to_check.extend(walk_source_dir(path))
+            elif path.is_file():
+                files_to_check.append(path)
+            else:
+                logger.error("Invalid path: %s", path)
+
+    # Use set() so we check each file only once.
+    reduced_file_list = sorted(set(files_to_check))
+
+    codefiles = []
+    for filename in reduced_file_list:
+        try:
+            # resolve() removes any '..' references from the path
+            path = Path(filename).resolve()
+            codefiles.append(TextFile.from_file(path))
+
+        except FileNotFoundError:
+            # resolve() may have failed. Use the input string.
+            logger.error("Could not open '%s'", filename)
+
+        except UnicodeDecodeError as ex:
+            logger.error(
+                "Failed to decode '%s' as %s (reason: %s, position: %d)",
+                filename,
+                ex.encoding,
+                ex.reason,
+                ex.start,
+            )
```

Comment:
This error handling is duplicated with `core.py`. Maybe they can be unified in some util function?